lib/expr.ex

```diff
@@ -3242,6 +3264,51 @@ defmodule AshSql.Expr do
     end
   end
 
+  # Returns a row-dependent token to prevent constant-folding for immutable functions, or nil if
+  # the repo hasn't opted-in to immutable expr_errors.
+  #
+  # The expression doesn't matter, as long as it depends on the row to avoid constant-folding.
```

Comment:
I worry that the expression needs to depend on the *full* primary key of the row actually, otherwise one row with a matching field value could be constant-folded to *not* raise an error.

Reply:
I guess in that case it wouldn't be constant folding, but it could just be cached.

Reply:
I don't think it _needs_ to based on my testing, but I actually did update this from that earlier version I showed you. This one concats all the primary key columns together in a text expression, falling back to the first attribute if there is no primary key.

I can drop the comment if you think it adds confusion?

Reply:
oh I just misread the code. Can we create an array instead of concatenating as a string? not all primary keys are guaranteed to support concatenation like that.

Reply:
Hm, I'm not sure an array would work since the primary key attributes could use different types. But how about building up a ROW of the primary keys to pass in like `ROW(D0."tenant_id", D0."row_id")`?